main/src/ca/uwaterloo/flix/language/ast/Ast.scala

```diff
@@ -408,20 +408,53 @@ object Ast {
 
     /**
       * Returns `this` labelled graph including only the edges where all
-      * its labels are in `syms` and the arity matches.
+      * its labels are in `syms` and the terms match.
       * A rule like
       * `A(ta) :- B(tb), not C(tc).` is represented by `edge(A, pos, {(A, arity(ta)), (B, arity(tb)), (C, arity(tc))}, (B, arity(tb)))` etc.
-      * and is only included in the output if `syms` contains all of `A, B, C` and `syms(A) == arity(ta)` etc.
+      * and is only included in the output if `syms` contains all of `A, B, C` and `maybeEq(syms(A), ta)` etc.
       */
-    def restrict(syms: Map[Name.Pred, Int]): LabelledGraph = {
+    def restrict(syms: Map[Name.Pred, Label]): LabelledGraph = {
+      def termsEq(t1: List[Type], t2: List[Type]): Boolean =
+        t1.length == t2.length &&
+          t1.zip(t2).forall{case (t1, t2) => maybeEq(t1, t2)}
+
       def check(l: Label): Boolean =
-        syms.get(l.pred).contains(l.arity)
+        syms.get(l.pred).exists(l0 => l0.lattice == l.lattice && termsEq(l0.terms, l.terms))
 
       LabelledGraph(edges.filter {
         case LabelledEdge(_, _, labels, _, _) =>
           labels.forall(check)
       })
     }
+
+    private val safeTypeConstructors = List(
+      TypeConstructor.Unit,
+      TypeConstructor.Null,
+      TypeConstructor.Bool,
+      TypeConstructor.Char,
+      TypeConstructor.Float32,
+      TypeConstructor.Float64,
+      TypeConstructor.Int8,
+      TypeConstructor.Int16,
+      TypeConstructor.Int32,
+      TypeConstructor.Int64,
+      TypeConstructor.BigInt,
+      TypeConstructor.Str,
+      TypeConstructor.Arrow,
+      TypeConstructor.Record,
+      TypeConstructor.Array,
+      TypeConstructor.Channel,
+      TypeConstructor.Lazy,
+      TypeConstructor.Native,
+      TypeConstructor.Tuple
+    )
+
+    private def maybeEq(t1: Type, t2: Type): Boolean = {
```

Comment:
You have to recurse on the structure of type. A type can have multiple type constructors for example `Option[List[...]]`.

Reply:
I would patmatch on (t1, t2) and then recursive through when they have the same structure.